geomstats/geometry/discrete_surfaces.py

```diff
@@ -320,8 +343,567 @@ def surface_metric_matrices(self, point):
+        vertex_areas_bp : array-like, shape=[n_vertices, 1]
+            Vertex areas for each vertex of the base_point.
+
+        Returns
+        -------
+        _ : array-like, shape=[...]
+            Term of order 2, and coefficient a2, of the inner-product.
+
+        References
+        ----------
+        .. [HSKCB2022] "Elastic shape analysis of surfaces with second-order
+            Sobolev metrics: a comprehensive numerical framework".
+            arXiv:2204.04238 [cs.CV], 25 Sep 2022.
+        """
+        laplacian_at_base_point = self._space.laplacian(base_point)
+        return self.a2 * gs.sum(
+            gs.einsum(
+                "...bi,...bi->...b",
+                laplacian_at_base_point(tangent_vec_a),
+                laplacian_at_base_point(tangent_vec_b),
+            )
+            / vertex_areas_bp,
+            axis=-1,
+        )
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        r"""Compute inner product between two tangent vectors at a base point.
+
+        The inner-product has 6 terms, where each term corresponds to
+        one of the 6 hyperparameters a0, a1, b1, c1, d1, a2.
+
+        We denote h and k the tangent vectors a and b respectively.
+        We denote q the base point, i.e. the surface.
+
+        The six terms of the inner-product are given by:
+        :math:`\int_M (G_{a_0} + G_{a_1} + G_{b_1} + G_{c_1} + G_{d_1} + G_{a_2})vol_q`
+
+        where:
+        - :math:`G_{a_0} = a_0 <h, k>`
+        - :math:`G_{a_1} = a_1.g_q^{-1} <dh_m, dk_m>`
+        - :math:`G_{b_1} = b_1.g_q^{-1} <dh_+, dk_+>`
+        - :math:`G_{c_1} = c_1.g_q^{-1} <dh_\perp, dk_\perp>`
+        - :math:`G_{d_1} = d_1.g_q^{-1} <dh_0, dk_0>`
+        - :math:`G_{a_2} = a_2 <\Delta_q h, \Delta_q k>`
+
+        with notations taken from .. [HSKCB2022].
+
+        Parameters
+        ----------
+        tangent_vec_a : array-like, shape=[..., n_vertices, 3]
+            Tangent vector at base point.
+        tangent_vec_b : array-like, shape=[..., n_vertices, 3]
+            Tangent vector at base point.
+        base_point : array-like, shape=[n_vertices, 3]
+            Surface, as the 3D coordinates of the vertices of its triangulation.
+
+        Returns
+        -------
+        inner_prod : array-like, shape=[...]
+            Inner-product.
+
+        References
+        ----------
+        .. [HSKCB2022] "Elastic shape analysis of surfaces with second-order
+            Sobolev metrics: a comprehensive numerical framework".
+            arXiv:2204.04238 [cs.CV], 25 Sep 2022.
+        """
+        to_squeeze = False
+        if tangent_vec_a.ndim == 2 and tangent_vec_b.ndim == 2:
+            to_squeeze = True
+        if tangent_vec_a.ndim == 2:
+            tangent_vec_a = gs.expand_dims(tangent_vec_a, axis=0)
+        if tangent_vec_b.ndim == 2:
+            tangent_vec_b = gs.expand_dims(tangent_vec_b, axis=0)
```

Comment:
I think this can be fully deleted after addressing suggestions in `_inner_product_d1` and modifying slightly this function.

in worst case scenario, `tangent_vec_a, tangent_vec_b= gs.broadcast_arrays(tangent_vec_a, tangent_vec_b)` makes this cleaner.

Reply:
I tried using:
```
tangent_vec_a, tangent_vec_b= gs.broadcast_arrays(tangent_vec_a, tangent_vec_b)
```
but it led to an error. Keeping the current version for now.